File: L4_reverse_DLL.cpp
#include<iostream>
#include<bits/stdc++.h>
using namespace std;

class Node{
public:
    int data;
    Node *prev;
    Node *next;

    Node(int x){
        data = x;
        prev = NULL;
        next = NULL;
    }
};

// 1 --> 2 --> 3 --> 4 --> 5 -->6


class Solution{
public:
    Node *reverseDLL(Node *head){
        if(head == NULL || head->next == NULL) return head;
        Node *lastNode = nullptr;
        Node *curr = head;

        while(curr){
            lastNode = curr->prev;
            curr->prev = curr->next;
            curr->next = lastNode;
            curr = curr->prev;
        }
        return lastNode->prev;
    }

    Node *arrToDLL(vector<int> &arr){
        int n = arr.size();
        Node *head = new Node(arr[0]);
        Node *prevNode = head;
        for(int i = 1; i<n; i++){
            Node *newNode = new Node(arr[i]);
            prevNode->next = newNode;
            newNode->prev = prevNode;
            prevNode = prevNode->next;
        }
        return head;
    } 
};

int main(){
    vector<int> arr = {1, 2, 3, 4, 5};
    Solution obj;

    cout<<"DLL: ";
    Node *head = obj.arrToDLL(arr);
    // Node *temp1 = head;
    // while(temp1){
    //     cout<<temp1->data<<" ";
    //     temp1 = temp1->next;
    // }

    Node *temp2 = obj.reverseDLL(head);
    while(temp2){
        cout<<temp2->data<<" ";
        temp2 = temp2->next;
    }
    return 0;
}

File: L4_reverse_DLL.java

 class Node {
     public int data;
     public Node next;
     public Node prev;

     Node()
     {
         this.data = 0;
         this.next = null;
         this.prev = null;
     }

     Node(int data)
     {
         this.data = data;
         this.next = null;
         this.prev = null;
     }

     Node(int data, Node next)
     {
         this.data = data;
         this.next = next;
         this.prev = next;
     }
 }

public class L4_reverse_DLL{
    public static Node reverseDLL(Node head){
        // Write your code here.
        if(head == null || head.next == null) return head;
         /*
         swap(a, b):=>swap(curr.prev, curr.next)
         temp = a; => lastNode = curr.prev;
         a = b;    => curr.prev =  curr.next;
         b = temp; => curr.next = lastNode
         */

        Node curr = head;
        Node lastNode = null;
        while(curr != null){
            lastNode = curr.prev;
            curr.prev = curr.next;
            curr.next = lastNode;
            curr = curr.prev;
        }
        return lastNode.prev;
    }
}
/*Node *reverseDLL(Node *head) {

  if (head == nullptr || head->next == nullptr)

    return head;

  Node *current = head;

  Node *last;

  while (current != nullptr) {

    last = current->prev;

    current->prev = current->next;

    current->next = last;

    current = current->prev;

  }

 

  return last->prev;

}*/